mParticle/mparticle-web-sdk
# feat: Add Support for Batch Persistence when Offline

## Instructions
 1. PR target branch should be against `development`
 2. PR title name should follow this format: https://github.com/mParticle/mparticle-workflows/blob/main/.github/workflows/pr-title-check.yml
 3. PR branch prefix should follow this format: https://github.com/mParticle/mparticle-workflows/blob/main/.github/workflows/pr-branch-check-name.yml

 ## Summary
 - Implements two instances of the Vault Local Storage Wrapper, one for Events, and a second for Batches
 - As events are queued, they are placed into local storage for later processing. Once the `prepareAndUpload` method is triggered, events are removed from local storage and converted to `Batches`, then placed into the batch vault.
 - Once batches are ready to upload, they are removed from local storage to prevent accidental retransmissions, and an upload to our servers is attempted.
 - If the transmission is successful, no further action is taken.
 - If the transmission is unsuccessful, the batches are added back to local storage and are retransmitted during the next interval.
 - This pull request also refactors parts of the BatchUploader to better handle asynchronous uploads. Currently, a single asynchronous request is made that iterates through a list of batches, yet each batch is individually uploaded to the server. Loops inside of promises are not recommended in JavaScript as they could lead a longer turnaround time to process the response or other unexpected results. By breaking up the single upload promise into individual promises, each upload is a single process that returns quicker requires less processing.
 - Also in this PR are updates to our event conversion logic so that we can make sure that consent state is not lost when deserializing events and batches when storing in local storage.
 - Lastly, Beacon Uploader tests are moved out of the Batch Uploader test suite as there was an issue with race condition in the test runner.

 ## Testing Plan
 - Use a sample app to manually test the event upload flow
 - Update your config to make `eventBatchingIntervalMillis` longer than `5000` milliseconds
 - Verify that events are stored in local storage and then removed and converted into batches
 - Verify that events upload as expected
 - Set browser to offline and verify that batches and events continue to get stored in local storage
 - Verify that batches are uploaded once connectivity is restored

 ## Reference Issue (For mParticle employees only.  Ignore if you are an outside contributor)
 - Closes https://go.mparticle.com/work/SQDSDKS-4715
